File: Sources/GatewayCore/SessionDefaults.swift
import Foundation

/// What the app renders with, until told otherwise.
///
/// **These are generation settings, not listening settings.** `AudioProfile`
/// holds how loud things play; this holds what gets made. They are kept apart
/// because one belongs in the render key and the other must never be.
///
/// It exists because the render voice was the string `"M1"`, written into
/// `RenderService` and again into `gfrender`. Auto-mode rendered with M1
/// whatever the user did, and retiring M1 broke the CLI's default — two copies
/// of a fact, both wrong at once, which is the shape of bug this codebase keeps
/// turning up.
public struct SessionDefaults: Codable, Sendable, Equatable {
    /// Empty means "whichever voice is ready" — resolved against the library at
    /// use, never guessed. A default naming a voice that may not exist is how
    /// the last one broke.
    public var voice: String = ""
    public var verbosity: Int = 3
    public var pauseScale: Double = 1.0

    public init() {}

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        voice = try c.decodeIfPresent(String.self, forKey: .voice) ?? ""
        verbosity = try c.decodeIfPresent(Int.self, forKey: .verbosity) ?? 3
        pauseScale = try c.decodeIfPresent(Double.self, forKey: .pauseScale) ?? 1.0
    }

    /// The voice to render with: the chosen one if it is still there and
    /// clonable, otherwise the first that is, otherwise nil.
    ///
    /// Falling back rather than failing matters because a voice can be retired
    /// between launches, and the alternative is an app that renders nothing and
    /// says only that "M1" is missing.
    public func resolvedVoice(in voices: [VoiceRef]) -> String? {
        resolution(in: voices).name
    }

    /// The same answer with its reasoning attached, for anything that needs to
    /// tell the listener a substitution happened. One rule, shared with
    /// templates, so the app default and a session plan cannot disagree about
    /// which voice is current.
    public func resolution(in voices: [VoiceRef]) -> VoiceResolution {
        // A saved default that exists but is not yet clonable is not honoured:
        // this value drives the render queue, and queueing work against a voice
        // that cannot clone would fail every take. A template's preference is
        // different -- it is read, not run -- so `VoiceResolution` keeps it.
        if !VoiceResolution.isUnspecified(voice),
           let v = voices.first(where: { $0.name == voice }), v.isClonable {
            return VoiceResolution(name: v.name, reason: .requested)
        }
        guard let fallback = VoiceResolution.best(in: voices) else {
            return VoiceResolution(name: nil, reason: .unavailable)
        }
        return VoiceResolution(
            name: fallback.name,
            reason: VoiceResolution.isUnspecified(voice)
                ? .unspecified : .substituted(requested: voice))
    }

    public var clampedVerbosity: Int { min(max(verbosity, 1), 3) }
    public var clampedPauseScale: Double {
        min(max(pauseScale, RenderPlan.pauseScaleRange.lowerBound),
            RenderPlan.pauseScaleRange.upperBound)
    }
}

public enum SessionDefaultsIO {
    public static func url(root: URL) -> URL { root.appending(path: "memory/session.json") }

    public static func load(root: URL) -> SessionDefaults {
        guard let d = try? Data(contentsOf: url(root: root)),
              let s = try? JSONDecoder().decode(SessionDefaults.self, from: d)
        else { return SessionDefaults() }
        return s
    }

    public static func save(_ s: SessionDefaults, root: URL) throws {
        let u = url(root: root)
        try FileManager.default.createDirectory(at: u.deletingLastPathComponent(),
                                                withIntermediateDirectories: true)
        let enc = JSONEncoder()
        enc.outputFormatting = [.prettyPrinted, .sortedKeys]
        try enc.encode(s).write(to: u, options: .atomic)
    }
}

File: cross-platform/src/core/sessionDefaults.ts
/**
 * What the app renders with, until told otherwise.
 *
 * **These are generation settings, not listening settings.** `AudioProfile`
 * holds how loud things play; this holds what gets made. They are kept apart
 * because one belongs in the render key and the other must never be.
 *
 * It exists because the render voice was the string `"M1"`, written into the
 * render service and again into the CLI. Auto-mode rendered with M1 whatever
 * the user did, and retiring M1 broke the CLI's default — two copies of a
 * fact, both wrong at once.
 */
import { pauseScaleRange } from "./renderPlan.js";
import {
  best, isUnspecified, resolveVoice,
  type ResolvableVoice, type VoiceResolution,
} from "./voiceResolution.js";

export interface SessionDefaults {
  /** Empty means "whichever voice is ready" — resolved against the library at
   *  use, never guessed. A default naming a voice that may not exist is how
   *  the last one broke. */
  voice: string;
  verbosity: number;
  pauseScale: number;
}

export const defaults = (): SessionDefaults => ({ voice: "", verbosity: 3, pauseScale: 1.0 });

export function decodeDefaults(raw: unknown): SessionDefaults {
  const d = defaults();
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return d;
  const o = raw as Record<string, unknown>;
  return {
    voice: typeof o.voice === "string" ? o.voice : d.voice,
    verbosity: typeof o.verbosity === "number" ? o.verbosity : d.verbosity,
    pauseScale: typeof o.pauseScale === "number" ? o.pauseScale : d.pauseScale,
  };
}

/**
 * The same answer as `resolveVoice` with its reasoning attached, but with one
 * deliberate difference: **a saved default that exists but cannot clone is not
 * honoured.** This value drives the render queue, and queueing work against a
 * voice that cannot clone would fail every take. A template's preference is
 * different — it is read, not run — so `resolveVoice` keeps it.
 */
export function resolution(d: SessionDefaults, voices: ResolvableVoice[]): VoiceResolution {
  if (!isUnspecified(d.voice)) {
    const v = voices.find(x => x.name === d.voice);
    if (v !== undefined && v.isClonable) return { name: v.name, reason: { kind: "requested" } };
  }
  const fallback = best(voices);
  if (fallback === undefined) return { reason: { kind: "unavailable" } };
  return {
    name: fallback.name,
    reason: isUnspecified(d.voice)
      ? { kind: "unspecified" }
      : { kind: "substituted", requested: d.voice },
  };
}

/** The voice to render with: the chosen one if it is still there and clonable,
 *  otherwise the first that is, otherwise undefined. Falling back rather than
 *  failing matters because a voice can be retired between launches. */
export const resolvedVoice = (d: SessionDefaults, voices: ResolvableVoice[]): string | undefined =>
  resolution(d, voices).name;

export const clampedVerbosity = (d: SessionDefaults): number =>
  Math.min(Math.max(d.verbosity, 1), 3);

export const clampedPauseScale = (d: SessionDefaults): number =>
  Math.min(Math.max(d.pauseScale, pauseScaleRange[0]), pauseScaleRange[1]);

// `resolveVoice` is re-exported so a caller comparing the two rules imports
// them from one place; the difference between them is the point.
export { resolveVoice };
